src/App.tsx

```diff
@@ -1,14 +1,77 @@
 /* eslint-disable max-len */
-import React from 'react';
+import React, { useCallback, useEffect, useRef, useState } from 'react';
 import 'bulma/css/bulma.css';
 import '@fortawesome/fontawesome-free/css/all.css';
 
 import { TodoList } from './components/TodoList';
 import { TodoFilter } from './components/TodoFilter';
 import { TodoModal } from './components/TodoModal';
 import { Loader } from './components/Loader';
+import { Todo } from './types/Todo';
+import { getTodos } from './api';
+import { Query } from './types/Query';
+
+const TODO_IS_ACTIVE = 'active';
+const NO_SPECIFIC_QUERY_FOR_TODO = 'all';
+
+const filterTodosByQuery = (
+  todos: Todo[],
+  { finishQuery, searchQuery }: Query,
+): Todo[] => {
+  return todos?.filter(todo => {
+    if (finishQuery === NO_SPECIFIC_QUERY_FOR_TODO) {
+      return todo.title.toLowerCase().includes(searchQuery.toLowerCase());
+    }
+
+    return (
+      todo.title.toLowerCase().includes(searchQuery.toLowerCase()) &&
+      todo.completed === (finishQuery === TODO_IS_ACTIVE ? false : true)
+    );
+  });
+};
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [areDataLoading, setAreDataLoading] = useState(true);
+  const initialTodos = useRef<Todo[]>([]);
+  const [selectedTodo, setSelectedTodo] = useState<Todo | null>(null);
+
+  useEffect(() => {
+    setAreDataLoading(true);
+    getTodos()
+      .then(todosList => {
+        setTodos(todosList);
+        initialTodos.current = todosList;
+      })
+      .catch(error => {
+        throw new Error(error);
+      })
+      .finally(() => {
+        setAreDataLoading(false);
+      });
+  }, []);
+
+  const handleEyeClick = useCallback(
+    (selectedTodoId: number) => {
+      setSelectedTodo(todos.find(todo => todo.id === selectedTodoId) || null);
+    },
+    [todos],
+  );
+
+  const handleFiltrationQueries = useCallback(
+    ({ finishQuery, searchQuery }: Query) => {
+      setTodos(
+        filterTodosByQuery(initialTodos.current, { finishQuery, searchQuery }),
+      );
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [initialTodos.current],
+  );
```

Comment:
As I said previously you don't need to store calculated values in the state, just in a variable. In state store all todos what you get from the API
Simple example:
```typescript
const filteredTodos = filterTodos(todos, { query, status });
```

Reply:
@BudnikOleksii 

okey look:

I receive `query `and `status` from handleFiltrationQueries, when user uses filtration fields, so fo relaying these parameters I'll have to create 2 new states in `App component` relay it to `TodoFilter component`, lift up from it and then relay `query `and `status` to `filterTodos`.
It looks not such flexible as I made it in my way. Moreover `filteredTodos `(in you suggestion) will be recreated every time and `filterTodos `will be recalculated every time, even if I don't change filtration fields.

Maybe I just haven't understood your solution. If that's the case, correct me.